pageserver/src/virtual_file.rs

```diff
@@ -666,33 +666,39 @@ impl VirtualFile {
                     offset += n as u64;
                 }
                 Err(e) if e.kind() == std::io::ErrorKind::Interrupted => {}
-                Err(e) => return (Slice::into_inner(buf), Err(e)),
+                Err(e) => return (restore(buf), Err(e)),
             }
         }
-        (Slice::into_inner(buf), Ok(()))
+        (restore(buf), Ok(()))
     }
 
-    /// Writes `buf.slice(0..buf.bytes_init())`.
-    /// Returns the IoBuf that is underlying the BoundedBuf `buf`.
-    /// I.e., the returned value's `bytes_init()` method returns something different than the `bytes_init()` that was passed in.
-    /// It's quite brittle and easy to mis-use, so, we return the size in the Ok() variant.
-    pub async fn write_all<B: BoundedBuf<Buf = Buf>, Buf: IoBuf + Send>(
+    /// Writes `buf` to the file at the current offset.
+    ///
+    /// Panics if there is an uninitialized range in `buf`, as that is most likely a bug in the caller.
+    pub async fn write_all<Buf: IoBuf + Send>(
         &mut self,
-        buf: B,
+        buf: Slice<Buf>,
         ctx: &RequestContext,
-    ) -> (B::Buf, Result<usize, Error>) {
+    ) -> (Slice<Buf>, Result<usize, Error>) {
+        assert_eq!(
+            buf.bytes_init(),
+            buf.bytes_total(),
+            "caller likely meant to fill the buffer fully"
+        );
+
         let nbytes = buf.bytes_init();
         if nbytes == 0 {
-            return (Slice::into_inner(buf.slice_full()), Ok(0));
+            return (buf, Ok(0));
         }
```

Comment:
Wouldn't this be covered by the `(Slice::from_buf_bounds(buf.into_inner(), bounds), Ok(nbytes))` on line 715?